File: cli/golem-cli/src/bridge_gen/type_naming/builder.rs
// Copyright 2024-2025 Golem Cloud
//
// Licensed under the Golem Source License v1.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://license.golem.cloud/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::bridge_gen::type_naming::type_location::{
    TypeLocation, TypeLocationPath, TypeLocationRoot,
};

pub(super) struct Builder {
    root_owner: RootOwner,
    root_item_name: String,
    path: Option<TypeLocationPath>,
}

impl Builder {
    pub fn new() -> Self {
        Self {
            root_owner: RootOwner::ConstructorInput,
            root_item_name: "".to_string(),
            path: None,
        }
    }

    pub fn set_root_owner(&mut self, owner: RootOwner) {
        self.root_owner = owner;
        self.root_item_name = "".to_string();
        self.path = None;
    }

    pub fn set_root_item_name(&mut self, name: &str) {
        self.root_item_name = name.to_string();
        self.path = None;
    }

    pub fn push(&mut self, elem: TypeLocationPath) {
        match &mut self.path {
            None => {
                self.path = Some(elem);
            }
            Some(path) => Self::push_path(path.inner_mut(), elem),
        }
    }

    fn push_path(path: &mut Option<Box<TypeLocationPath>>, elem: TypeLocationPath) {
        match path {
            Some(path) => {
                Self::push_path(path.inner_mut(), elem);
            }
            None => {
                *path = Some(Box::new(elem));
            }
        }
    }

    pub fn pop(&mut self) {
        if let Some(path) = &mut self.path {
            if !Self::pop_path(path) {
                self.path = None;
            }
        }
    }

    fn pop_path(path: &mut TypeLocationPath) -> bool {
        enum InnerPopResult {
            NoInner,
            Popped,
            NotPopped,
        }

        let inner_pop_result = match path.inner_mut() {
            Some(inner) => {
                if Self::pop_path(inner) {
                    InnerPopResult::Popped
                } else {
                    InnerPopResult::NotPopped
                }
            }
            None => InnerPopResult::NoInner,
        };

        match inner_pop_result {
            InnerPopResult::NoInner => false,
            InnerPopResult::Popped => true,
            InnerPopResult::NotPopped => {
                *path.inner_mut() = None;
                true
            }
        }
    }

    pub fn type_location(&self) -> TypeLocation {
        TypeLocation {
            root: match &self.root_owner {
                RootOwner::ConstructorInput => TypeLocationRoot::AgentConstructorInput {
                    input_name: self.root_item_name.clone(),
                },
                RootOwner::MethodInput { method_name } => TypeLocationRoot::AgentMethodInput {
                    method_name: method_name.clone(),
                    input_name: self.root_item_name.clone(),
                },
                RootOwner::MethodOutput { method_name } => TypeLocationRoot::AgentMethodOutput {
                    method_name: method_name.clone(),
                    output_name: self.root_item_name.clone(),
                },
            },
            path: self.path.clone(),
        }
    }
}

pub(super) enum RootOwner {
    ConstructorInput,
    MethodInput { method_name: String },
    MethodOutput { method_name: String },
}

File: sdks/ts/packages/golem-ts-sdk/src/internal/schema-model/builder.ts
// Copyright 2024-2026 Golem Cloud
//
// Licensed under the Golem Source License v1.1 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://license.golem.cloud/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Schema-graph construction helpers.
//
// `SchemaBuilder` registers named (nominal) types into a graph using a
// reserve/commit protocol so recursive and mutually-recursive types close
// without infinite recursion: a type reserves its id before walking its body,
// so a back-reference encountered during the walk short-circuits to a `ref`.
//
// `mergeAgentGraphs` combines the per-root graphs of an agent's constructor,
// methods, and config into a single graph, deduplicating shared definitions by
// stable `type-id` and rejecting divergent definitions of the same id.

import {
  type SchemaGraph,
  type SchemaType,
  type SchemaTypeDef,
  type TypeId,
  deepEqual,
  emptyMetadata,
} from './model';
import { SchemaConflictError } from './errors';

interface ReservedDef {
  name?: string;
  body?: SchemaType;
}

export class SchemaBuilder {
  private readonly defs = new Map<TypeId, ReservedDef>();

  /** Whether the given id is already reserved or committed. */
  contains(id: TypeId): boolean {
    return this.defs.has(id);
  }

  /** Reserve a slot for `id` so recursive references can close to a `ref`. */
  reserve(id: TypeId, name?: string): void {
    if (!this.defs.has(id)) {
      this.defs.set(id, { name, body: undefined });
    }
  }

  /** Commit the body of a previously reserved (or new) `id`. */
  commit(id: TypeId, body: SchemaType, name?: string): void {
    const existing = this.defs.get(id);
    if (existing) {
      existing.body = body;
      if (name !== undefined) existing.name = name;
    } else {
      this.defs.set(id, { name, body });
    }
  }

  /** A `ref` schema type pointing at `id`. */
  ref(id: TypeId): SchemaType {
    return { body: { tag: 'ref', id }, metadata: emptyMetadata() };
  }

  /**
   * Register a nominal type. If `id` is not yet known it is reserved, then
   * `build` is invoked to produce its body (during which recursive references
   * to `id` resolve to a `ref`), then committed. Always returns a `ref` to
   * `id` for use at the call site.
   */
  register(id: TypeId, build: () => SchemaType, name?: string): SchemaType {
    if (!this.contains(id)) {
      this.reserve(id, name);
      this.commit(id, build(), name);
    }
    return this.ref(id);
  }

  /** Finalize the registered definitions, ensuring every reservation was committed. */
  finish(): Map<TypeId, SchemaTypeDef> {
    const out = new Map<TypeId, SchemaTypeDef>();
    for (const [id, def] of this.defs) {
      if (def.body === undefined) {
        throw new Error(`schema builder: reserved type id '${id}' was never committed`);
      }
      out.set(id, { name: def.name, body: def.body });
    }
    return out;
  }

  /** Build a complete graph with the given root and all registered definitions. */
  buildGraph(root: SchemaType): SchemaGraph {
    return { defs: this.finish(), root };
  }
}

/** Merge definitions from several graphs, rejecting conflicting same-id bodies. */
export function mergeGraphDefs(graphs: Iterable<SchemaGraph>): Map<TypeId, SchemaTypeDef> {
  const merged = new Map<TypeId, SchemaTypeDef>();
  for (const g of graphs) {
    for (const [id, def] of g.defs) {
      const existing = merged.get(id);
      if (existing) {
        if (!deepEqual(existing.body, def.body)) {
          throw new SchemaConflictError(id);
        }
      } else {
        merged.set(id, def);
      }
    }
  }
  return merged;
}

/** An agent-level merged graph: a shared def registry plus the roots it backs. */
export interface MergedAgentGraph {
  defs: Map<TypeId, SchemaTypeDef>;
  roots: SchemaType[];
}

/**
 * Combine the per-root graphs of an agent (one per constructor / method input /
 * output / config root) into a single deduplicated def registry plus the list
 * of roots, in input order. Conflicting same-id definitions raise
 * `SchemaConflictError`.
 */
export function mergeAgentGraphs(graphs: SchemaGraph[]): MergedAgentGraph {
  const defs = mergeGraphDefs(graphs);
  const roots = graphs.map((g) => g.root);
  return { defs, roots };
}
